### ISSUE TITLE
Add RtMidi and RtAudio backends to support other OSs

### ISSUE THREAD

**Issue by dagargo** [dagargo/elektroid#65]

**Comment by snd1** [comment]
https://www.music.mcgill.ca/~gary/rtaudio/

**Comment by dagargo** [comment]
MIDI support thru RtMidi is ready. See d4bf2bbd50a92d2ed0c1fcd43584d90724d0ade1.

In Linux, ALSA is used for MIDI by default. In order to use RtMidi under Linux, `RTMIDI=yes` needs to be passed to `./configure`.

**Comment by dagargo** [comment]
Audio support thru RtAudio is ready. See 162869f63515917cf46b7dd60e6ed199df29947f.

In Linux, similarly to RtMidi, PulseAudio is used for audio by default. In order to use RtAudio under Linux, `RTAUDIO=yes` needs to be passed to ./configure.

Of course, both options can be passed simultaneously, which makes Elektroid OS independent a least theoretically.

I haven't tested it yet neither on OSX nor on Windows. In case anyone is interested, we'll need to properly test these for the next release.

**Comment by dagargo** [comment]
Porting Elektroid to OSX is gonna be a little more complicated than I naively assumed.

So far, these are the packages installed with `brew install`.

* automake
* libtool
* gtk+3
* libsndfile
* libsamplerate
* libzip
* rtmidi
* rtaudio
* glib
* json-glib

But there are some issues.

* In brew, RtMidi 5 is used by default. Debian stable uses version 4 but testing (Bookworm) uses version 5. Version should be upgraded.
* The `inotify` code should be changed to use [fswatch[(https://github.com/emcrisostomo/fswatch), which is a cross platform library that provides the same functionality.
* `ld` reports some undefined symbols for endianness functions and `strverscmp`.

```
Undefined symbols for architecture x86_64:
  "_be16toh", referenced from:
      _elektron_copy_sample_data in elektroid-elektron.o
      _elektron_next_data_entry in elektroid-elektron.o
  "_be32toh", referenced from:
      _elektron_next_smplrw_entry in elektroid-elektron.o
      _elektron_download_smplrw in elektroid-elektron.o
      _elektron_upload_smplrw in elektroid-elektron.o
      _elektron_next_data_entry in elektroid-elektron.o
      _elektron_download_data_prefix in elektroid-elektron.o
      _elektron_open_datum in elektroid-elektron.o
      _elektron_close_datum in elektroid-elektron.o
      ...
  "_be64toh", referenced from:
      _elektron_get_storage_stats in elektroid-elektron.o
  "_htobe16", referenced from:
      _elektron_tx_and_rx_timeout_no_cache in elektroid-elektron.o
      _elektron_new_msg_write_sample_blk in elektroid-elektron.o
  "_htobe32", referenced from:
      _elektron_upgrade_os in elektroid-elektron.o
      _elektron_get_sample_path_from_hash_size in elektroid-elektron.o
      _elektron_new_msg_read_sample_blk in elektroid-elektron.o
      _elektron_new_msg_close_sample_read in elektroid-elektron.o
      _elektron_new_msg_write_sample_blk in elektroid-elektron.o
      _elektron_new_msg_close_sample_write in elektroid-elektron.o
      _elektron_new_msg_open_common_write in elektroid-elektron.o
      ...
  "_le32toh", referenced from:
      _sample_load_raw in elektroid-sample.o
  "_strverscmp", referenced from:
      _sample_get_sample_extensions in elektroid-sample.o
```

I'm leaving this open until Audio and MIDI can be tested.

**Comment by dagargo** [comment]
I've upgraded RtMidi version to 5.0 in 196b6b9722913a1c089d51224f2398bf37c1fdf1.

Notice that the upcoming Debian 12 includes this version but Debian 11 does not. So RtMidi support has been dropped on Debian 11.

This shouldn't be a problem as RtMidi should used only on non Linux Systems.

**Comment by snd1** [comment]
No worry, new versions usually don't go to older debian versions, unless someone backports this.

Debian has GNU/Hurd(1), as well as GNU/kFreeBSD(2), but not as official archs/kernels.

Also, the first stage of the freeze for Debian 12 started already on 12th January(3)

1: https://www.debian.org/ports/hurd/
2: https://www.debian.org/ports/kfreebsd-gnu/
3: https://release.debian.org/testing/freeze_policy.html

**Comment by dagargo** [comment]
@snd1, do you think version 2.4 could make it into Debian12? I was aware of the timeline and I've just re-read the policy but it's still not clear to me. I guess "no large/disruptive changes" do not apply here. :thinking: 

The idea is to not make any more changes and release version 2.4 during the upcoming week.

**Comment by snd1** [comment]
No hurry... This means if i can upload 2.4 before 2023-02-12 (Soft Freeze), then it can be in main archive. If not, then i can backport 2.4 to bullseye :)

**Comment by snd1** [comment]
@dagargo did you or anybody tested rtmidi and rtaudio support and/or it is already fully implemented in 2.4.1? i haven't the time to test this, or better said no machine atm with gnu/hurd or kfreebsd, sorry. So i wonder if i should enable hurd and kfreebsd again.

**Comment by dagargo** [comment]
RtMidi and RtAudio are fully working on 2.4.1. At least, I've tested these against a VM running Debian 12 running Linux. Notice RtMidi dependency is version 5.0 so it will only work on Debian 12.

I haven't installed Hurd or kfreebsd either so I suggest against enabling them for now.

As this issue is completed, I'm gonna close it. But I've created  #86 and #87 to keep track of the tests over kfreebsd and Hurd.
